### PR TITLE
Spmddebugchecknov26

### PR DESCRIPTION
**Pull request by ManfeiBai**

### CODE REVIEW

File: torch_xla/distributed/spmd/__init__.py [hunk 1]
```diff
@@ -3,10 +3,21 @@
...
+    "ShardingType",
+    "ShardingSpec",
+    "XLAPatchedLinear",
+    "mark_sharding",
+    "clear_sharding",
+    "wrap_if_sharded",
+    "xla_distribute_tensor",
+    "xla_distribute_module",
+    "xla_patched_nn_linear_forward",
+    "visualize_tensor_sharding",
```
Comment (yeounoh) [inline torch_xla/distributed/spmd/__init__.py:22]: Let's not directly expose, but under `torch_xla.distributed.spmd.debugging.visualize_tensor_sharding`

File: test/spmd/test_spmd_debugging.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+import io
+import rich
+
+import torch
+import torch_xla
+import torch_xla.runtime as xr
+import torch_xla.utils.utils as xu
+import torch_xla.core.xla_env_vars as xenv
+import torch_xla.core.xla_model as xm
+import torch_xla.experimental.xla_sharding as xs
```
Comment (yeounoh) [inline test/spmd/test_spmd_debugging.py:17]: use `import torch_xla.distributed.spmd as xs` instead. Same thing with the other SPMD related imports.

File: test/spmd/test_spmd_debugging.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+class DebuggingSpmdTest(test_xla_sharding_base.XlaShardingTest):
+
+  @classmethod
+  def setUpClass(cls):
+    xr.use_spmd()
+    super().setUpClass()
+
+  @unittest.skipIf(
+      not xr.using_pjrt() or
+      xu.getenv_as(xenv.PJRT_DEVICE, str) in ("GPU", 'CUDA', 'ROCM'),
```
Comment (yeounoh) [inline test/spmd/test_spmd_debugging.py:33]: Check if it's TPU or CPU instead.

File: test/spmd/test_spmd_debugging.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+      f"Requires PJRT_DEVICE set to `TPU`, `CPU`.")
+  def test_debugging_spmd_single_host_tiled(self):
+    from torch_xla.distributed.spmd.debugging import visualize_tensor_sharding
+    device = xm.xla_device()
+    num_devices = xr.global_runtime_device_count()
```
Comment (yeounoh) [inline test/spmd/test_spmd_debugging.py:38]: You can just use `self.n_devices`, `self.device_ids` instead. No need to call & construct them again.

File: test/spmd/test_spmd_debugging.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    mesh_shape = (2, num_devices // 2)
+    device_ids = np.array(range(num_devices))
+    mesh = Mesh(device_ids, mesh_shape, ('x', 'y'))
```
Comment (yeounoh) [inline test/spmd/test_spmd_debugging.py:41]: You can just use `mesh = self._get_mesh(mesh_shape)`, and skip `device_ids`

File: test/spmd/test_spmd_debugging.py [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+    t = torch.randn(8, 4, device=device)
+    partition_spec = (0, 1)
+    xs.mark_sharding(t, mesh, partition_spec)
+    sharding = torch_xla._XLAC._get_xla_sharding_spec(t)
+    generatedtable = visualize_tensor_sharding(t)
+    console = rich.console.Console(file=io.StringIO(), width=120)
+    console.print(generatedtable)
```
Comment (yeounoh) [inline test/spmd/test_spmd_debugging.py:48]: Do not write /print to terminal. `output = console.file.gevalue()` should be sufficient for teting.

File: test/spmd/test_spmd_debugging.py [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+    fask_table.add_row(*col)
+    fake_console.print(fask_table)
+    fake_output = fake_console.file.getvalue()
+    assert output == fake_output
+
+  @unittest.skipIf(
+      not xr.using_pjrt() or
+      xu.getenv_as(xenv.PJRT_DEVICE, str) in ("GPU", 'CUDA', 'ROCM'),
+      f"Requires PJRT_DEVICE set to `TPU`, 'CPU'.")
+  def test_single_host_partial_replication(self):
```
Comment (yeounoh) [inline test/spmd/test_spmd_debugging.py:113]: ditto.

File: test/spmd/test_spmd_debugging.py [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+    fake_console = rich.console.Console(file=io.StringIO(), width=120)
+    fake_console.print(fask_table)
+    fake_output = fake_console.file.getvalue()
+    assert output == fake_output
+
+  @unittest.skipIf(
+      not xr.using_pjrt() or
+      xu.getenv_as(xenv.PJRT_DEVICE, str) in ("GPU", 'CUDA', 'ROCM'),
+      f"Requires PJRT_DEVICE set to `TPU`, 'CPU'.")
+  def test_single_host_replicated(self):
```
Comment (yeounoh) [inline test/spmd/test_spmd_debugging.py:163]: ditto

**Comment by ManfeiBai** [comment]
since https://github.com/pytorch/xla/pull/5742 passed CI, will close this test PR

**Pull request closed without merging**